// ng-parts/src/app/modules/device/device.service.ts
import {Injectable} from '@angular/core';
import {Subject} from 'rxjs/Subject';

// Device
import {Device} from './device';

@Injectable()
export class DeviceService {

  private device: Device = new Device();

  private _device = new Subject<Device>();
  device$ = this._device.asObservable();

  public breakPoints: any = {
    mobile: '(max-width: 767px)',
    tablet: '(min-width: 768px) and (max-width: 1023px)',
    desktop: '(min-width: 1024px)'
  };

  constructor() {
    this.setFamily();

    let resizeEnd;
    window.addEventListener('resize', () => {
      clearTimeout(resizeEnd);
      resizeEnd = setTimeout(() => {
        this.setFamily();
      }, 200);
    }, false);
  }

  setFamily(): void {
    if (window.matchMedia(this.breakPoints.mobile).matches) {
      this.device.family = 'mobile';
    } else if (window.matchMedia(this.breakPoints.tablet).matches) {
      this.device.family = 'tablet';
    } else if (window.matchMedia(this.breakPoints.desktop).matches) {
      this.device.family = 'desktop';
    }

    this._device.next(this.device);
  }

  get Device(): Device {
    return this.device;
  }
}
